R/ivreg.marg.R

```diff
@@ -0,0 +1,160 @@
+#' @export
+#' @rdname marg
+#' @examples
+#'
+#' # --------------------------------------
+#' # 2SLS using AER::ivreg (example adapted from ?AER::ivreg)
+#'
+#' data(margex)
+#' margex$assign <- margex$treatment
+#'
+#' # Probability of getting treatment increases with age and distance
+#' margex$pr_treat <- plogis(
+#'   margex$age / sd(margex$age) +
+#'   margex$distance / sd(margex$distance))
+#'
+#' # One-way non-compliance
+#' margex$actual <- margex$assign
+#' margex$actual[margex$assign == 1] <- rbinom(
+#'   n = sum(margex$assign == 1), size = 1,
+#'   prob = margex$pr_treat[margex$assign == 1]
+#' )
+#'
+#' mod <- AER::ivreg(y ~ as.factor(actual) + age + distance |
+#'                     as.factor(assign) + age + distance,
+#'                   data = margex)
+#'
+#'
+#' v <- function(object, ...){
+#'   vcov(object) * object$df.residual / nobs(object)
+#' }
+#'
+#' summary(mod, vcov = v)
+#'
+#' # Get the level of the outcome variable at different values of `actual`
+#' # Note data is a *required* parameter here
+#' marg(mod, var_interest = 'actual', data = margex, type = 'levels')
+#' marg(mod, var_interest = 'actual', data = margex, type = 'effects')
+#'
+marg.ivreg <- function(mod, var_interest,
+                       type = 'levels',
+                       vcov_mat = NULL,
+                       dof = NULL,
+                       at = NULL, base_rn = 1,
+                       at_var_interest = NULL,
+                       data,
+                       weights = mod$weights,
+                       cofint = 0.95){
+  # Set params and Run checks ----
+
+  check_inputs(weights = weights, data = data, var_interest = var_interest,
+               at = at, cofint = cofint, base_rn = base_rn, type = type,
+               dof = dof, vcov_mat = vcov_mat)
+
+  # Checks specific to ivreg ----
+  # See if we're looking for continuous variables
+  if(type == 'effects' & is.numeric(data[[var_interest]]) &
+     ! all(unique(data[[var_interest]]) %in% c(0, 1)) &
+     ! sprintf("as.character(%s)", var_interest) %in% names(data) &
+     ! sprintf("as.factor(%s)", var_interest) %in% names(data))
+    stop('We do not support effects for continuous variables at this time.')
+
+  # Check if no weights when model was built was weights
+  if(is.null(weights) & !is.null(mod$weights))
+    warning('The model was built with weights, but you have not ',
+            'provided weights. Your calculated margins may be odd. ',
+            'See Details.')
+
+  # Keep only complete variables ----
+  data <- clean_glm_data(mod, data, weights)
```

Comment:
I agree - I started feeling that way as I was modifying the function; I expected to need to tear up a lot more than was necessary. I do think there's probably a better way to structure marg.

But... should that be a different MR?